Add C_AppendString C function for appending to a string of type TAG_STRING. Can append multiple items to a string.

skittles1 opened the pull request:
This C function takes a string of type TAG_STRING or $ for the first parameter, and any number of following parameters and appends them in order to the string. If passed $, creates a new string. Returns the int value of the old or created string.

Can accept data of type TAG_INT, TAG_STRING, TAG_DEBUGSTR, TAG_MESSAGE, TAG_TEMP_STRING, TAG_RESOURCE and notably, TAG_LIST. If passed a list, will add each element of the list in order (if valid data) to the list. This means a list containing objects (which we don't want) and strings (which we do) would have just the strings added (currently this would log an error message, but shows one possible use).

Usage:

```
new_string = AppendString($, "data", " ", "for our new string");
```

skittles1 committed:
Add C_AppendString C function for adding to TAG_STRING.

This C function takes a string of type TAG_STRING or $ for the first
parameter, and any number of following parameters and appends them in
order to the string. If passed $, creates a new string. Returns the int
value of the old or created string.

Can accept data of type TAG_INT, TAG_STRING, TAG_DEBUGSTR, TAG_MESSAGE,
TAG_TEMP_STRING, TAG_RESOURCE and notably, TAG_LIST. If passed a list,
will add each element of the list in order (if valid data) to the list.
This means a list containing objects (which we don't want) and strings
(which we do) would have just the strings added.

treymd committed:
Linux compatibility for blakserv updates Fixed incompatible network code,
split off linux specific header file, Added replacements for windows
specific functions, defined more common windows types Implemented pthread versions of mutex functions

treymd committed:
Fixed bof loading for linux

treymd committed:
Make blakcomp build on linux

treymd committed:
Build kod tree on linux

treymd committed:
IPv6 for linux

treymd committed:
More work on the kod tree build

treymd committed:
Fixed portability issues in server BSP functions

treymd committed:
Added windows style thread message queue

treymd committed:
Switched to Andrew Kirmse's function porting method

treymd committed:
move thread messaging calls to separate files

treymd committed:
Split MainServer() and MainServerExit into separate files.  Split interface into separate files for windows and linux

treymd commented:
needs to be named concat, that is the accepted name in just about every language.

treymd committed:
implemented linux CLI

treymd committed:
Implemented async socket accept and reads for linux

treymd committed:
fixed minot error with mutexes
fixed resource case for some roo files
Fixed a few bof loading issues
Added some boilerplate text for some new source files
fixed filename case of jgeneral.kod

skittles1 committed:
Game time related code refactoring.

Renamed the game time related code in system.kod from NewHour etc. to
NewGameHour. Called messages that deal with game time as opposed to real
time now include Game in the message name. This was done to separate
them from future calls from RealTime that also use NewHour etc. but
refer to real hours vs. game ones.

System's RecalcLightAndWeather message renamed to SystemInitGameHour to
better describe what it actually does (setting initial game hour based
on GetTime()).

Renamed System's NewGameHour to SysRecalcLightAndDayPhase to better
describe what it does.

Added calls from System's time messages to GameEventEngine to update it
on game time.

RealTime's NewMinute message will now trigger RealTime's NewHour call on
the hour and so on.

skittles1 committed:
Fix rare message parameter related compiler bug.

Duplicate parameters were being checked using the current message ID for
comparison, however as the parameters are parsed before the message
header, these message IDs referred to the previous message.

This works fine nearly all the time (and the message ID isn't used for
anything else by the parameters) however if a set of parameters is being
parsed and the previous message in that class has a copy in another
(parsed in this batch) class followed by a message containing a
parameter with the same name, the compiler will return a false positive
for a duplicate parameter.

This change removes the duplicate check during parameter parsing and
moves it to the parameter list sorting.

skittles1 committed:
Fix errors in pre/post increment/decrement handling.

The existing code worked for simple calculations however trying to use
the prefix/postfix operators in complex statements where the result
(from a calc on a property) needs to be stored in a temp local var, the
mismatch here between source and dest types was causing errors.

Briefly, prefix ops should perform the operation then if the source is
not the destination, store the result in both places (opcode.dest gives
dest type, source1 gives source type, opnode->source gives source ID,
opnode->dest gives dest ID). Postfix ops should store the initial value
in the destination if not equal to the source location, perform the op
then store the result in the source location (basically a cheaper +/-1).

All existing operator usage works (have to use it on a property as part
of a larger calculation to trigger the error).

skittles1 committed:
Add /**/ (multi-line) and // style comments to kod.

We will potentially be using Doxygen or other codedoc program to
generate documentation from the codebase, and generating useful docs
will require the ability to comment kod code with multi-line comments.

The % style comments have been left intact, and support for both //
single line and /* */ multi-line comments has been added. Neither are
active inside strings, otherwise comments can be placed anywhere (source
is compiled as if they are not there at all).

Multi-line comments can be nested (making /* /* */ */ a valid comment)
but // takes precedence and can 'comment out' an /* inside a comment
block.

Also changed yyerror to exit as continuing to parse after this kind of
error can cause the compiler to crash.

Commit:
Set attack-request interval to 250ms

TheDiggie committed:
swaped out some useless items for good ones

treymd committed:
Add text file for CC 3.0 attributions

skittles1 committed:
Remove unnecessary makefiles variables in kod build.

BOFS2 is only used once, 3-8 not at all. I'm guessing at some point in
the past the number of files to one nmake variable was limited, or
perhaps they were used for neater formatting in the makefiles. Either
way, they aren't needed.

skittles1 committed:
Cleanup makefiles, fix a few rsc references.

Removed some missing bof references from makefiles and added some
(commented) bofs from kods that are present but not used.

Fixed some incorrect resources in a couple files.

skittles1 committed:
Fix compiler suggested issues

Fixed some issues suggested by VS's code analysis. The memset issue in
synched.c was also recently fixed in the upstream repo.

skittles1 committed:
Fix overlay drawing for new helm.

This handles a missing case in the overlay system - overlays that should
be drawn on top of a base item, but underneath underlays drawn on other
overlays on the base item.

Software renderer handles this by drawing the new type of overlay before
any other overlays are drawn on the base item, hardware does the same
but also has to set a Z bias equal to that of the base item to avoid
rendering the overlay in the wrong place.

Reproduced the kod hotspot index constants in object3d.h, HS_HELM is
used in FindHotspotPdib rather than using '2'.

skittles1 committed:
Add #region and #endregion folding.

Allow folding regions in editors that support it. The = and ; characters
are disallowed in region naming to avoid conflicting with #parameter=x
in Send/Create.

treymd committed:
- add code to fork non-interactive startup to the background.
- interactive startup option is blakserv -i (this will give you a command line interface to the server)

treymd committed:
Restore mysql settings to default config file

treymd committed:
Update VS project file to handle new OS dependent source files

Daenks committed:
Merge branch 'mfix' of https://github.com/skittles1/Meridian59_103 into build.1.0.1283-req.1337-br.mfix-auth.skittles1

Daenks committed:
Merge branch 'kodbuildfixes' of https://github.com/skittles1/Meridian59_103 into build.1.0.1283-req.1336-br.kodbuildfixes-auth.skittles1

Daenks committed:
Merge branch 'chest' of https://github.com/TheDiggie/Meridian59_103 into build.1.0.1283-req.1335-br.chest-auth.TheDiggie

Daenks committed:
Merge branch 'credits' of https://github.com/treymd/Meridian59_103 into build.1.0.1283-req.1333-br.credits-auth.treymd

Daenks committed:
Merge branch 'reqinterval' of https://github.com/cyberjunk/meridian59-orig into build.1.0.1283-req.1332-br.reqinterval-auth.cyberjunk

Daenks committed:
Merge branch 'multicomment' of https://github.com/skittles1/Meridian59_103 into build.1.0.1283-req.1331-br.multicomment-auth.skittles1

Daenks committed:
Merge branch 'compilefix' of https://github.com/skittles1/Meridian59_103 into build.1.0.1283-req.1322-br.compilefix-auth.skittles1

Daenks committed:
Merge branch 'linux' of https://github.com/treymd/Meridian59_103 into build.1.0.1283-req.1228-br.linux-auth.treymd

Daenks committed:
Merge branch 'master' into develop

skittles1 committed:
Add regions/doxygen comments in questengine.kod.

Added examples of region folding and doxygen-style comments in
questengine.kod. Some file modifications still necessary for doxygen to
parse it.

treymd committed:
Makefile addition from ckirmse 101 PR #338, silence complaints from creation of build directory that already exists

Daenks committed:
Merge branch 'master' into develop

skittles1 committed:
Remove patcher from repo, moved to separate repo.

The patcher has its own repository now, and this version is outdated.

skittles1 committed:
Add CONTRIBUTING.md with PR/coding standards.

Add PR/commit/coding standards file to provide some guidance for
contributors.

skittles1 committed:
Change linked-list of timers into min binary heap.

Timers are currently kept in a linked list, unfortunately the timers
generally have a cluster at the beginning of the list (mob movement)
so every insert/delete has to traverse minimum 30-100 elements before
potentially finding the correct place to modify.

Changing to min-heap means that worst-case scenario, 10 timers have
to be looked at (generally it is much better than this).

On my system, CreateTimer drops from 1us to 0.350us, DeleteTimer
from 1.1us to 0.6us and GetTimeRemaining from 8.6us to 1.5us.

skittles1 committed:
Use struct/union in timer_node to move data.

Benchmarks faster than moving elements individually (about 100ns faster
for CreateTimer).

skittles1 committed:
Add comments and region to timer.c.

skittles1 committed:
Added timer mem realloc and timer heap check function.

Timer heap memory will now be expanded if it hits the maximum allocated
timer nodes (default 2000).

Added a new menu option to the admin menu, Check, which currently has
one option, timerheap. "Check timerheap" will traverse the timer heap
and check it for validity (children of a node larger than parent). The
result and time to run is displayed in admin window/server interface.

Daenks committed:
Merge branch 'contrib' of https://github.com/skittles1/Meridian59_103 into build.1.0.1286-req.1346-br.contrib-auth.skittles1

Daenks committed:
Merge branch 'rempatcher' of https://github.com/skittles1/Meridian59_103 into build.1.0.1286-req.1345-br.rempatcher-auth.skittles1

Daenks committed:
Merge branch '101_ckirmse_338' of https://github.com/treymd/Meridian59_103 into build.1.0.1286-req.1342-br.101_ckirmse_338-auth.treymd

Daenks committed:
Merge branch 'kodcomment' of https://github.com/skittles1/Meridian59_103 into build.1.0.1286-req.1341-br.kodcomment-auth.skittles1

Daenks committed:
Merge branch 'newday' of https://github.com/skittles1/Meridian59_103 into build.1.0.1286-req.1321-br.newday-auth.skittles1

skittles1 commented:
Directed to master instead of develop, will change name to some variant of StringConcat, retest and resubmit later.
